processor/deltatocumulativeprocessor/internal/streams/expiry.go

```diff
@@ -0,0 +1,60 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package streams // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/internal/streams"
+
+import (
+	"time"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/exp/metrics/identity"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/exp/metrics/staleness"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/exp/metrics/streams"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/internal/clock"
+)
+
+func ExpireAfter[T any](items streams.Map[T], ttl time.Duration) Expiry[T] {
+	exp := Expiry[T]{
+		Staleness: *staleness.NewStaleness[T](ttl, items),
+		sig:       make(chan struct{}),
+	}
+	return exp
+}
+
+var _ streams.Map[any] = (*Expiry[any])(nil)
+
+type Expiry[T any] struct {
+	staleness.Staleness[T]
+	sig chan struct{}
+}
+
+func (e Expiry[T]) ExpireOldEntries() <-chan struct{} {
+	e.Staleness.ExpireOldEntries()
+
+	n := e.Staleness.Len()
+	sig := make(chan struct{})
+
+	go func() {
+		switch {
+		case n == 0:
+			<-e.sig
+		case n > 0:
+			expires := e.Staleness.Next().Add(e.Max)
+			until := clock.Until(expires)
+			clock.Sleep(until)
+		}
+		close(sig)
+	}()
+	return sig
+}
+
+func (e Expiry[T]) Store(id identity.Stream, v T) error {
+	err := e.Staleness.Store(id, v)
+
+	// "try-send" to notify possibly sleeping expiry routine
+	select {
+	case e.sig <- struct{}{}:
```

Comment:
This means that for every store, we're potentially starting a new "sleep" go-routine for expiry. Lots of stores quickly in a row could build up many go-routines that are all sleeping. Since each loop of creates a new one.

Reply:
are you sure?

afaict there is only one routine (started in Processor.Start). the linked code here is a try-send operation that either sends an empty struct into the channel (if the other routine happens to listen) or just does nothing at all

Reply:
In processor.Start() we launch a goroutine that does this:

```
for {
	p.mtx.Lock()
	next := p.exp.ExpireOldEntries()
	p.mtx.Unlock()

	select {
	case <-next:
	case <-p.ctx.Done():
		return
	}
}
```

`ExpireOldEntries()` is overridden above. It starts a goroutine to do the sleep. That's the goroutine I'm talking about. 

By sending a struct to the channel, you cause `next` to return early. IE, it doesn't wait for the sleep to finish / the channel to be closed. Which is the behaviour you *want*, but it also abandons that goroutine. The code will immediately loop and call `ExpireOldEntries()` again, which spawns a new goroutine. Etc.